File: src/AetherNet.Transport/Bandwidth/BandwidthEstimator.cs
// SPDX-License-Identifier: MIT

using System.Collections.Concurrent;
using AetherNet.Bandwidth;

namespace AetherNet.Transport.Bandwidth;

/// <summary>
/// Reference BBRv3-inspired bandwidth estimator for a single transport link.
///
/// <para>
/// Algorithm summary:
/// <list type="bullet">
///   <item><b>BtlBw (bottleneck bandwidth):</b> rolling maximum of per-delivery delivery-rate
///     samples over a <see cref="BtlBwWindowSize"/> = 10-RTprop window. Using the maximum (not
///     average) ensures we track the pipe capacity, not the current load. This mirrors the
///     BBRv3 BtlBwFilter (draft-cardwell-iccrg-bbr-congestion-control-02 §4.3.2.1).</item>
///   <item><b>RTprop (path propagation delay):</b> rolling minimum RTT over a
///     <see cref="RtPropWindowMs"/> = 10 000 ms window. The minimum filters out queueing delay.
///     Periodically forces a ProbeRTT phase to keep the estimate fresh.</item>
///   <item><b>SRTT / RTTVAR:</b> RFC 6298 §2.3 Jacobson/Karels algorithm. α = 1/8, β = 1/4.
///     Used for RTO and for confidence-tier promotion.</item>
///   <item><b>Loss rate:</b> EWMA with α = <see cref="LossAlpha"/> = 0.10.</item>
///   <item><b>PHY cap:</b> RSSI-to-BtlBw mapping constrains the estimate on weak radio links
///     before probe data arrives. Calibration tables are from IEEE 802.11 / 3GPP TS 36.213.</item>
/// </list>
/// </para>
///
/// <para>Thread safety: all state is protected by a single <c>lock</c>. Read-only
/// properties return volatile copies so readers never block on the lock.</para>
/// </summary>
public sealed class BandwidthEstimator : IBandwidthEstimator
{
    // ── Constants ────────────────────────────────────────────────────────────

    /// <summary>Number of delivery-rate samples kept in the BtlBw max-filter window.</summary>
    public const int BtlBwWindowSize = 10;

    /// <summary>Minimum RTT window duration in milliseconds (BBRv3 ProbeRTT period).</summary>
    public const double RtPropWindowMs = 10_000.0;

    /// <summary>EWMA loss rate smoothing factor (α).</summary>
    public const double LossAlpha = 0.10;

    /// <summary>RFC 6298 SRTT smoothing factor (1/8).</summary>
    private const double SrttAlpha = 0.125;

    /// <summary>RFC 6298 RTTVAR smoothing factor (1/4).</summary>
    private const double RttVarBeta = 0.25;

    /// <summary>5% improvement threshold for SampleImproved event.</summary>
    private const double ImprovementThreshold = 0.05;

    // ── State ────────────────────────────────────────────────────────────────

    private readonly object _lock = new();

    // BtlBw max-filter: circular buffer of (deliveryRateBps, timestampMs) samples.
    private readonly (long rateBps, double timestampMs)[] _btlBwWindow =
        new (long, double)[BtlBwWindowSize];
    private int _btlBwHead;
    private int _btlBwCount;

    // RTprop min-filter: (rttMs, timestampMs) pairs kept for the window duration.
    private readonly Queue<(double rttMs, double timestampMs)> _rtPropSamples = new();

    // RFC 6298 SRTT / RTTVAR
    private double _srttMs;
    private double _rttVarMs;
    private bool _firstRtt = true;

    // Loss EWMA
    private double _lossRate;

    // PHY cap
    private long _phyCapBps;

    // Confidence
    private int _probeRounds;

    // Snapshot cache (updated after every observation)
    private volatile BandwidthSample _current;

    // Gossip warm-start flag
    private bool _warmedFromGossip;

    // ── Constructor ──────────────────────────────────────────────────────────

    public BandwidthEstimator(string transportName, long maxBandwidthBps)
    {
        TransportName = transportName;
        // Optimistic initialisation: start at theoretical max with None confidence.
        // PHY hints and probes will tighten this quickly.
        _current = BuildSnapshot(maxBandwidthBps, TimeSpan.FromMilliseconds(50));
    }

    // ── IBandwidthEstimator ─────────────────────────────────────────────────

    public string TransportName { get; }

    public long BtlBwBps => _current.BtlBwBps;
    public long AvailableBps => _current.AvailableBps;
    public long BdpBytes => _current.BdpBytes;
    public TimeSpan Srtt => _current.Srtt;
    public TimeSpan RttVar => _current.RttVar;
    public TimeSpan RtProp => _current.RtProp;
    public double LossRate => _current.LossRate;
    public BandwidthConfidence Confidence => _current.Confidence;
    public BandwidthSample CurrentSample => _current;

    public event EventHandler<BandwidthSample>? SampleImproved;

    // ── Observation feed ─────────────────────────────────────────────────────

    public void RecordDelivery(int bytes, long sendTimestampUs, long deliverTimestampUs)
    {
        if (bytes <= 0 || deliverTimestampUs <= sendTimestampUs) return;

        var elapsedMs = (deliverTimestampUs - sendTimestampUs) / 1000.0;
        var deliveryRateBps = (long)(bytes * 8.0 / (elapsedMs / 1000.0));
        var rttMs = elapsedMs; // one-way → treat as RTT estimate (conservative)

        lock (_lock)
        {
            AddToBtlBwWindow(deliveryRateBps, NowMs());
            UpdateRttEstimates(rttMs);
            _probeRounds++;
            Commit();
        }
    }

    public void RecordLoss(int bytes)
    {
        if (bytes <= 0) return;
        lock (_lock)
        {
            _lossRate = LossAlpha * 1.0 + (1 - LossAlpha) * _lossRate;
            Commit();
        }
    }

    public void RecordProbeResult(BandwidthProbeAck ack, long localReceiveUs)
    {
        var rtt = ack.Rtt;
        if (rtt <= TimeSpan.Zero || rtt > TimeSpan.FromSeconds(30)) return;

        // Delivery rate from probe: bytes × 8 / RTT
        var deliveryRateBps = ack.ProbeBytes > 0
            ? (long)(ack.ProbeBytes * 8.0 / rtt.TotalSeconds)
            : 0L;

        lock (_lock)
        {
            UpdateRttEstimates(rtt.TotalMilliseconds);
            if (deliveryRateBps > 0)
                AddToBtlBwWindow(deliveryRateBps, NowMs());
            _probeRounds++;
            Commit();
        }
    }

    public void WarmFromGossip(long btlBwBps, TimeSpan rtProp, BandwidthConfidence sourceConfidence)
    {
        lock (_lock)
        {
            if (_probeRounds > 0 || _warmedFromGossip) return; // never downgrade
            // Seed one BtlBw window sample and RTprop estimate.
            AddToBtlBwWindow(btlBwBps, NowMs());
            var rttMs = rtProp.TotalMilliseconds;
            if (rttMs > 0)
            {
                _srttMs = rttMs;
                _rttVarMs = rttMs / 2.0;
                _firstRtt = false;
                AddToRtPropWindow(rttMs, NowMs());
            }
            _warmedFromGossip = true;
            Commit();
        }
    }

    public void ApplyPhyHint(int rssiDbm)
    {
        // RSSI → theoretical capacity mapping.
        // BLE calibration from Bluetooth SIG Core Spec 5.4 Table 7.2 (2Msym/s PHY):
        //   ≥ -70 dBm → up to  2 000 kbps
        //   ≥ -85 dBm → up to    500 kbps
        //   ≥ -95 dBm → up to    125 kbps
        //   < -95 dBm → up to     40 kbps (marginal link)
        // Wi-Fi (802.11ax) calibration from 3GPP TS 36.213 Annex A:
        //   ≥ -50 dBm → up to 600 Mbps
        //   ≥ -67 dBm → up to 200 Mbps
        //   ≥ -80 dBm → up to  54 Mbps
        //   < -80 dBm → up to  11 Mbps
        // We use the BLE table as a conservative fallback since all transports
        // share this interface; callers should use transport-specific mappings.
        long cap = rssiDbm switch
        {
            >= -50 => 600_000_000L,
            >= -67 =>  200_000_000L,
            >= -70 =>    2_000_000L,
            >= -80 =>   54_000_000L,
            >= -85 =>      500_000L,
            >= -95 =>      125_000L,
            _      =>       40_000L,
        };

        lock (_lock)
        {
            _phyCapBps = cap;
            Commit();
        }
    }

    // ── Internal helpers ─────────────────────────────────────────────────────

    /// <summary>
    /// RFC 6298 §2.3 RTT sample integration.
    /// First sample initialises SRTT = R, RTTVAR = R/2.
    /// Subsequent: RTTVAR = (1-β)×RTTVAR + β×|SRTT-R|; SRTT = (1-α)×SRTT + α×R.
    /// </summary>
    private void UpdateRttEstimates(double rttMs)
    {
        if (_firstRtt)
        {
            _srttMs = rttMs;
            _rttVarMs = rttMs / 2.0;
            _firstRtt = false;
        }
        else
        {
            _rttVarMs = (1 - RttVarBeta) * _rttVarMs + RttVarBeta * Math.Abs(_srttMs - rttMs);
            _srttMs = (1 - SrttAlpha) * _srttMs + SrttAlpha * rttMs;
        }

        // Success sample → also update loss EWMA (0 loss observed).
        _lossRate = LossAlpha * 0.0 + (1 - LossAlpha) * _lossRate;

        AddToRtPropWindow(rttMs, NowMs());
    }

    /// <summary>
    /// Insert a delivery-rate sample into the max-filter window.
    /// Discards old samples whose timestamp is outside 10×RTprop.
    /// </summary>
    private void AddToBtlBwWindow(long rateBps, double nowMs)
    {
        // Evict samples older than 10×RTprop.
        var windowDurationMs = 10.0 * Math.Max(1.0, MinRtPropMs());
        var expiry = nowMs - windowDurationMs;
        // Walk from tail and drop expired entries (circular buffer cleanup).
        while (_btlBwCount > 0)
        {
            var tail = (_btlBwHead + BtlBwWindowSize - _btlBwCount) % BtlBwWindowSize;
            if (_btlBwWindow[tail].timestampMs < expiry)
                _btlBwCount--;
            else
                break;
        }

        // Add new sample.
        _btlBwWindow[_btlBwHead] = (rateBps, nowMs);
        _btlBwHead = (_btlBwHead + 1) % BtlBwWindowSize;
        if (_btlBwCount < BtlBwWindowSize) _btlBwCount++;
    }

    private void AddToRtPropWindow(double rttMs, double nowMs)
    {
        _rtPropSamples.Enqueue((rttMs, nowMs));
        // Evict samples older than RtPropWindowMs.
        while (_rtPropSamples.Count > 0 && _rtPropSamples.Peek().timestampMs < nowMs - RtPropWindowMs)
            _rtPropSamples.Dequeue();
    }

    private long MaxBtlBwBps()
    {
        if (_btlBwCount == 0) return 0L;
        var max = 0L;
        for (var i = 0; i < _btlBwCount; i++)
        {
            var idx = (_btlBwHead + BtlBwWindowSize - _btlBwCount + i) % BtlBwWindowSize;
            if (_btlBwWindow[idx].rateBps > max) max = _btlBwWindow[idx].rateBps;
        }
        return max;
    }

    private double MinRtPropMs()
    {
        if (_rtPropSamples.Count == 0) return _srttMs > 0 ? _srttMs : 50.0;
        var min = double.MaxValue;
        foreach (var (rttMs, _) in _rtPropSamples)
            if (rttMs < min) min = rttMs;
        return min > 0 ? min : 1.0;
    }

    private BandwidthConfidence ComputeConfidence() => _probeRounds switch
    {
        0 when !_warmedFromGossip => BandwidthConfidence.None,
        0 => BandwidthConfidence.Low,
        < 5 => BandwidthConfidence.Low,
        < 20 => BandwidthConfidence.Medium,
        _ => BandwidthConfidence.High,
    };

    /// <summary>Rebuild the snapshot and fire SampleImproved if significant.</summary>
    private void Commit()
    {
        var prev = _current;
        _current = BuildSnapshot(MaxBtlBwBps(), TimeSpan.FromMilliseconds(MinRtPropMs()));
        var cur = _current;

        // Fire SampleImproved if BtlBw improved by ≥5% or confidence tier advanced.
        if (prev.BtlBwBps == 0 ||
            (cur.BtlBwBps - prev.BtlBwBps) > prev.BtlBwBps * ImprovementThreshold ||
            cur.Confidence > prev.Confidence)
        {
            // Fire outside the lock to avoid deadlocks.
            ThreadPool.QueueUserWorkItem(_ => SampleImproved?.Invoke(this, cur));
        }
    }

    private BandwidthSample BuildSnapshot(long btlBw, TimeSpan rtProp)
    {
        var srtt = TimeSpan.FromMilliseconds(Math.Max(1.0, _srttMs));
        var rttVar = TimeSpan.FromMilliseconds(Math.Max(0.0, _rttVarMs));
        var lossClamp = Math.Clamp(_lossRate, 0.0, 1.0);
        // Effective = the PHY-capped deliverable rate. BDP and AvailableBps are
        // both derived from the EFFECTIVE rate — the BDP must size the in-flight
        // window to the rate the link can actually carry, not the uncapped BtlBw.
        var effective = _phyCapBps > 0 ? Math.Min(btlBw, _phyCapBps) : btlBw;
        var bdp = effective > 0 ? (long)(effective / 8.0 * rtProp.TotalSeconds) : 0L;

        return new BandwidthSample(
            TransportName,
            effective,          // BtlBwBps capped by PHY hint
            (long)(effective * (1.0 - lossClamp)),
            bdp,
            srtt,
            rttVar,
            rtProp,
            _lossRate,
            _phyCapBps,
            ComputeConfidence(),
            DateTimeOffset.UtcNow);
    }

    private static double NowMs() =>
        DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
}

File: typescript/src/bandwidth/BandwidthEstimator.ts
// SPDX-License-Identifier: MIT

/**
 * AetherNet Bandwidth Measurement Framework — per-transport estimator.
 *
 * Ported from:
 *   src/AetherNet.Transport/Bandwidth/BandwidthEstimator.cs
 *
 * BBRv3-inspired four-phase state machine.  JavaScript is single-threaded so
 * no mutex is needed — just plain object mutation.
 */

import {
  BandwidthConfidence,
  type BandwidthSample,
  type BandwidthProbeAck,
  makeBandwidthSample,
} from "./models.js";

// ── Constants ─────────────────────────────────────────────────────────────────

/** Number of delivery-rate samples kept in the BtlBw max-filter window. */
export const BTL_BW_WINDOW_SIZE = 10;

/** Minimum RTT window duration in milliseconds (BBRv3 ProbeRTT period). */
export const RT_PROP_WINDOW_MS = 10_000.0;

/** EWMA loss rate smoothing factor (α). */
export const LOSS_ALPHA = 0.10;

/** RFC 6298 SRTT smoothing factor (1/8). */
const SRTT_ALPHA = 0.125;

/** RFC 6298 RTTVAR smoothing factor (1/4). */
const RTT_VAR_BETA = 0.25;

/** 5 % improvement threshold for onSampleImproved callbacks. */
const IMPROVEMENT_THRESHOLD = 0.05;

// ── BandwidthEstimator ────────────────────────────────────────────────────────

/**
 * Per-transport link bandwidth estimator.
 *
 * Two observation paths:
 *   Passive — recordDelivery() feeds the BBRv3 delivery-rate filter.
 *   Active  — recordProbeResult() consumes ack timestamps to maintain
 *             estimates on idle links.
 *
 * AetherNet innovations:
 *   PHY-layer capping — applyPhyHint() constrains estimates before probes.
 *   Gossip warm-start — warmFromGossip() pre-seeds from a peer's value.
 *   Confidence tiers  — consumers can distinguish 1-probe from 30-round.
 */
export class BandwidthEstimator {
  // ── BtlBw max-filter: circular buffer of (rateBps, timestampMs) ──────────
  private readonly _btlBwWindow: Array<{ rateBps: bigint; timestampMs: number }>;
  private _btlBwHead = 0;
  private _btlBwCount = 0;

  // ── RTprop min-filter ────────────────────────────────────────────────────
  private readonly _rtPropSamples: Array<{ rttMs: number; timestampMs: number }> = [];

  // ── RFC 6298 SRTT / RTTVAR ───────────────────────────────────────────────
  private _srttMs = 0.0;
  private _rttVarMs = 0.0;
  private _firstRtt = true;

  // ── Loss EWMA ────────────────────────────────────────────────────────────
  private _lossRate = 0.0;

  // ── PHY cap ──────────────────────────────────────────────────────────────
  private _phyCapBps: bigint = 0n;

  // ── Confidence ───────────────────────────────────────────────────────────
  private _probeRounds = 0;

  // ── Snapshot cache ────────────────────────────────────────────────────────
  private _current: BandwidthSample;

  // ── Gossip warm-start flag ────────────────────────────────────────────────
  private _warmedFromGossip = false;

  /**
   * Callbacks invoked when BtlBw improves by ≥ 5 % or confidence advances.
   * Equivalent to the C# SampleImproved event.
   */
  public onSampleImproved: Array<(sample: BandwidthSample) => void> = [];

  constructor(
    public readonly transportName: string,
    private readonly _maxBandwidthBps: bigint,
  ) {
    this._btlBwWindow = Array.from({ length: BTL_BW_WINDOW_SIZE }, () => ({
      rateBps: 0n,
      timestampMs: 0,
    }));
    // Optimistic initialisation: start at theoretical max with None confidence.
    this._current = this._buildSnapshot(_maxBandwidthBps, 50.0);
  }

  // ── Public accessors ──────────────────────────────────────────────────────

  get currentSample(): BandwidthSample {
    return this._current;
  }

  // ── Observation feed ──────────────────────────────────────────────────────

  /**
   * Record a successful delivery of `bytes`.
   * Both timestamps are microseconds since Unix epoch on the same clock.
   */
  recordDelivery(bytes: number, sendUs: bigint, deliverUs: bigint): void {
    if (bytes <= 0 || deliverUs <= sendUs) return;

    const elapsedMs = Number(deliverUs - sendUs) / 1000.0;
    const deliveryRateBps = BigInt(Math.round((bytes * 8.0) / (elapsedMs / 1000.0)));
    const rttMs = elapsedMs; // one-way → treat as RTT estimate (conservative)

    this._addToBtlBwWindow(deliveryRateBps, this._nowMs());
    this._updateRttEstimates(rttMs);
    this._probeRounds++;
    this._commit();
  }

  /** Record that `bytes` were lost (timeout or explicit NAK). */
  recordLoss(bytes: number): void {
    if (bytes <= 0) return;
    this._lossRate = LOSS_ALPHA * 1.0 + (1 - LOSS_ALPHA) * this._lossRate;
    this._commit();
  }

  /**
   * Feed an active probe ack into the estimator.
   * `localReceiveUs` is the local clock µs at ACK receipt (currently unused
   * in the algorithm — kept for API symmetry with the C# reference).
   */
  recordProbeResult(ack: BandwidthProbeAck, localReceiveUs: bigint): void {
    void localReceiveUs; // intentionally unused; kept for API symmetry
    const rttMs = ack.rtt;
    if (rttMs <= 0 || rttMs > 30_000) return;

    // Delivery rate from probe: bytes × 8 / RTT
    const deliveryRateBps =
      ack.probeBytes > 0
        ? BigInt(Math.round((ack.probeBytes * 8.0) / (rttMs / 1000.0)))
        : 0n;

    this._updateRttEstimates(rttMs);
    if (deliveryRateBps > 0n) {
      this._addToBtlBwWindow(deliveryRateBps, this._nowMs());
    }
    this._probeRounds++;
    this._commit();
  }

  /**
   * Pre-warm from a gossip payload.
   * Only effective when confidence is None — never downgrades an existing estimate.
   * `rtPropMs` is milliseconds.
   */
  warmFromGossip(
    btlBwBps: bigint,
    rtPropMs: number,
    confidence: BandwidthConfidence,
  ): void {
    void confidence; // used for guard logic via _probeRounds / _warmedFromGossip
    if (this._probeRounds > 0 || this._warmedFromGossip) return; // never downgrade

    const now = this._nowMs();
    this._addToBtlBwWindow(btlBwBps, now);
    if (rtPropMs > 0) {
      this._srttMs = rtPropMs;
      this._rttVarMs = rtPropMs / 2.0;
      this._firstRtt = false;
      this._addToRtPropWindow(rtPropMs, now);
    }
    this._warmedFromGossip = true;
    this._commit();
  }

  /**
   * Apply a physical-layer hint.
   * RSSI-to-BtlBw caps the estimate before probes complete.
   * Uses the BLE calibration table as a conservative fallback.
   */
  applyPhyHint(rssiDbm: number): void {
    let cap: bigint;
    if      (rssiDbm >= -50) cap = 600_000_000n;
    else if (rssiDbm >= -67) cap = 200_000_000n;
    else if (rssiDbm >= -70) cap =   2_000_000n;
    else if (rssiDbm >= -80) cap =  54_000_000n;
    else if (rssiDbm >= -85) cap =     500_000n;
    else if (rssiDbm >= -95) cap =     125_000n;
    else                     cap =      40_000n;

    this._phyCapBps = cap;
    this._commit();
  }

  // ── Internal helpers ──────────────────────────────────────────────────────

  /**
   * RFC 6298 §2.3 RTT sample integration.
   * First sample initialises SRTT = R, RTTVAR = R/2.
   */
  private _updateRttEstimates(rttMs: number): void {
    if (this._firstRtt) {
      this._srttMs = rttMs;
      this._rttVarMs = rttMs / 2.0;
      this._firstRtt = false;
    } else {
      this._rttVarMs =
        (1 - RTT_VAR_BETA) * this._rttVarMs +
        RTT_VAR_BETA * Math.abs(this._srttMs - rttMs);
      this._srttMs = (1 - SRTT_ALPHA) * this._srttMs + SRTT_ALPHA * rttMs;
    }

    // Success sample → also update loss EWMA (0 loss observed).
    this._lossRate = LOSS_ALPHA * 0.0 + (1 - LOSS_ALPHA) * this._lossRate;

    this._addToRtPropWindow(rttMs, this._nowMs());
  }

  /**
   * Insert a delivery-rate sample into the max-filter window.
   * Discards samples older than 10×RTprop.
   */
  private _addToBtlBwWindow(rateBps: bigint, nowMs: number): void {
    const windowDurationMs = 10.0 * Math.max(1.0, this._minRtPropMs());
    const expiry = nowMs - windowDurationMs;

    // Evict expired entries from the tail of the circular buffer.
    while (this._btlBwCount > 0) {
      const tail =
        (this._btlBwHead + BTL_BW_WINDOW_SIZE - this._btlBwCount) %
        BTL_BW_WINDOW_SIZE;
      if (this._btlBwWindow[tail].timestampMs < expiry) {
        this._btlBwCount--;
      } else {
        break;
      }
    }

    this._btlBwWindow[this._btlBwHead] = { rateBps, timestampMs: nowMs };
    this._btlBwHead = (this._btlBwHead + 1) % BTL_BW_WINDOW_SIZE;
    if (this._btlBwCount < BTL_BW_WINDOW_SIZE) this._btlBwCount++;
  }

  private _addToRtPropWindow(rttMs: number, nowMs: number): void {
    this._rtPropSamples.push({ rttMs, timestampMs: nowMs });
    // Evict samples older than RT_PROP_WINDOW_MS.
    const expiry = nowMs - RT_PROP_WINDOW_MS;
    let removeCount = 0;
    for (const s of this._rtPropSamples) {
      if (s.timestampMs < expiry) removeCount++;
      else break;
    }
    if (removeCount > 0) this._rtPropSamples.splice(0, removeCount);
  }

  private _maxBtlBwBps(): bigint {
    if (this._btlBwCount === 0) return 0n;
    let max = 0n;
    for (let i = 0; i < this._btlBwCount; i++) {
      const idx =
        (this._btlBwHead + BTL_BW_WINDOW_SIZE - this._btlBwCount + i) %
        BTL_BW_WINDOW_SIZE;
      if (this._btlBwWindow[idx].rateBps > max)
        max = this._btlBwWindow[idx].rateBps;
    }
    return max;
  }

  private _minRtPropMs(): number {
    if (this._rtPropSamples.length === 0)
      return this._srttMs > 0 ? this._srttMs : 50.0;
    let min = Number.MAX_VALUE;
    for (const s of this._rtPropSamples) {
      if (s.rttMs < min) min = s.rttMs;
    }
    return min > 0 ? min : 1.0;
  }

  private _computeConfidence(): BandwidthConfidence {
    if (this._probeRounds === 0 && !this._warmedFromGossip)
      return BandwidthConfidence.None;
    if (this._probeRounds === 0) return BandwidthConfidence.Low;
    if (this._probeRounds < 5)  return BandwidthConfidence.Low;
    if (this._probeRounds < 20) return BandwidthConfidence.Medium;
    return BandwidthConfidence.High;
  }

  /** Rebuild the snapshot and fire onSampleImproved if significant. */
  private _commit(): void {
    const prev = this._current;
    this._current = this._buildSnapshot(this._maxBtlBwBps(), this._minRtPropMs());
    const cur = this._current;

    const prevBtl = prev.btlBwBps;
    const improved =
      prevBtl === 0n ||
      cur.btlBwBps - prevBtl > BigInt(Math.round(Number(prevBtl) * IMPROVEMENT_THRESHOLD)) ||
      cur.confidence > prev.confidence;

    if (improved && this.onSampleImproved.length > 0) {
      // Fire asynchronously (matches C# ThreadPool.QueueUserWorkItem behaviour).
      const snap = cur;
      Promise.resolve().then(() => {
        for (const cb of this.onSampleImproved) cb(snap);
      });
    }
  }

  private _buildSnapshot(btlBw: bigint, rtPropMs: number): BandwidthSample {
    const srttMs  = Math.max(1.0, this._srttMs);
    const rttVarMs = Math.max(0.0, this._rttVarMs);
    const lossClamp = Math.min(Math.max(this._lossRate, 0.0), 1.0);
    const available = BigInt(Math.round(Number(btlBw) * (1.0 - lossClamp)));
    const bdpBytes  = btlBw > 0n
      ? BigInt(Math.round(Number(btlBw) / 8.0 * (rtPropMs / 1000.0)))
      : 0n;
    const effective = this._phyCapBps > 0n
      ? (btlBw < this._phyCapBps ? btlBw : this._phyCapBps)
      : btlBw;
    const effectiveAvailable = BigInt(Math.round(Number(effective) * (1.0 - lossClamp)));

    return makeBandwidthSample({
      transportName: this.transportName,
      btlBwBps:      effective,
      availableBps:  effectiveAvailable,
      bdpBytes,
      srttMs,
      rttVarMs,
      rtPropMs,
      lossRate:      this._lossRate,
      phyCapBps:     this._phyCapBps,
      confidence:    this._computeConfidence(),
      measuredAt:    new Date(),
    });
  }

  private _nowMs(): number {
    return Date.now();
  }
}
